### PR TITLE
Check Overflow Cases in Calculating VarListSize

### PR DESCRIPTION
**Pull request by os-d**

### CODE REVIEW

File: SetupDataPkg/Include/Library/ConfigVariableListLib.h [hunk 1]
```diff
@@ -49,6 +46,25 @@ typedef struct {
 } CONFIG_VAR_LIST_HDR;
 #pragma pack(pop)
 
+/**
+  Return the size of the variable list given a NameSize (including null terminator) and DataSize
+
+  @param[in]  NameSize    Size of the CHAR16 name of the config knob including null terminator
```
Comment (kuqin12) [inline SetupDataPkg/Include/Library/ConfigVariableListLib.h]: Probably want to mention this is in the unit of bytes?

File: SetupDataPkg/Library/ConfigVariableListLib/ConfigVariableListLib.c [hunk 2]
```diff
@@ -14,6 +14,48 @@
 #include <Library/DxeServicesLib.h>
 #include <Library/PcdLib.h>
 #include <Library/ConfigVariableListLib.h>
+#include <Library/SafeIntLib.h>
+
+/**
+  Return the size of the variable list given a NameSize (including null terminator) and DataSize
+
+  @param[in]  NameSize    Size of the CHAR16 name of the config knob including null terminator
```
Comment (kuqin12) [inline SetupDataPkg/Library/ConfigVariableListLib/ConfigVariableListLib.c]: Same here?

File: SetupDataPkg/Library/ConfigVariableListLib/ConfigVariableListLib.c [hunk 3]
```diff
@@ -14,6 +14,48 @@
...
+  UINT32         CalculatedSize = sizeof (CONFIG_VAR_LIST_HDR) + sizeof (EFI_GUID) + sizeof (UINT32) + sizeof (UINT32);
+  RETURN_STATUS  Status;
+
+  if (NeededSize == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  Status = SafeUint32Add (CalculatedSize, NameSize, &CalculatedSize);
+  if (RETURN_ERROR (Status)) {
+    return EFI_BUFFER_TOO_SMALL;
```
Comment (kuqin12) [inline SetupDataPkg/Library/ConfigVariableListLib/ConfigVariableListLib.c]: just return the status?

File: SetupDataPkg/Library/ConfigVariableListLib/ConfigVariableListLib.c [hunk 4]
```diff
@@ -66,20 +108,26 @@ ConvertVariableListToVariableEntry (
...
-  BinSize    = *Size;
-  VarList    = (CONST CONFIG_VAR_LIST_HDR *)((CHAR8 *)VariableListBuffer);
-  NeededSize = VAR_LIST_SIZE (VarList->NameSize, VarList->DataSize);
+  BinSize = *Size;
+  VarList = (CONST CONFIG_VAR_LIST_HDR *)((CHAR8 *)VariableListBuffer);
+  Status  = GetVarListSize (VarList->NameSize, VarList->DataSize, &NeededSize);
+
+  if (EFI_ERROR (Status)) {
+    // we overflowed
+    DEBUG ((DEBUG_ERROR, "%a VarList size overflowed, too large of config!\n", __FUNCTION__));
```
Comment (kuqin12) [inline SetupDataPkg/Library/ConfigVariableListLib/ConfigVariableListLib.c]: Print the 2 sizes?

File: SetupDataPkg/Library/ConfigVariableListLib/ConfigVariableListLib.c [hunk 5]
```diff
@@ -185,13 +233,24 @@ ConvertVariableEntryToVariableList (
     goto Exit;
   }
 
-  NameSize = StrnSizeS (VariableEntry->Name, CONF_VAR_NAME_LEN);
+  Status = SafeUintnToUint32 (StrnSizeS (VariableEntry->Name, CONF_VAR_NAME_LEN), &NameSize);
+  if (EFI_ERROR (Status)) {
+    // overflowed...
+    DEBUG ((DEBUG_ERROR, "%a VarListEntry name %a overflowed, too long a name!\n", __FUNCTION__, VariableEntry->Name));
+    goto Exit;
```
Comment (kuqin12) [inline SetupDataPkg/Library/ConfigVariableListLib/ConfigVariableListLib.c]: nit: Given this is bound to `CONF_VAR_NAME_LEN`, i think we do not have to use the safe int here.

But if it really overflow, we probably do not want to print the bogus name to run into illegal access.

**Review by kuqin12 — APPROVED** [review]

File: SetupDataPkg/Library/ConfigVariableListLib/ConfigVariableListLib.c (hunk 3 above)
Comment (os-d) [reply SetupDataPkg/Library/ConfigVariableListLib/ConfigVariableListLib.c]: I know that RETURN_STATUS == EFI_STATUS, but I didn't know if we should rely on that assumption (I'm sure everyone everywhere does...).

File: SetupDataPkg/Library/ConfigVariableListLib/ConfigVariableListLib.c (hunk 5 above)
Comment (os-d) [reply SetupDataPkg/Library/ConfigVariableListLib/ConfigVariableListLib.c]: Agreed, wasn't paying attention and got overzealous with the SafeIntLib :). I'll remove this usage.

**Pull request merged**